### PR TITLE
1556 backupstore errors

### PR DESCRIPTION
**Pull request by boknowswiki**
This PR covers three fixes:

1. Fix potential data race
2. Fix restore backup incrementally test case
3. Fix nfs.go go fmt error

https://github.com/longhorn/longhorn/issues/1556

### CODE REVIEW

**Comment by yasker** [comment]
There are seems race condition in the `FileLock`. @joshimoo 
```
WARNING: DATA RACE
Write at 0x00c0000c23e0 by goroutine 34:
  github.com/longhorn/backupstore.(*FileLock).Unlock()
      /go/src/github.com/longhorn/backupstore/lock.go:126 +0x102
  github.com/longhorn/backupstore.CreateDeltaBlockBackup.func1()
      /go/src/github.com/longhorn/backupstore/deltablock.go:230 +0x2cd

Previous read at 0x00c0000c23e0 by goroutine 32:
  github.com/longhorn/backupstore.(*FileLock).Refresh.func1()
      /go/src/github.com/longhorn/backupstore/lock.go:149 +0xb5

Goroutine 34 (running) created at:
  github.com/longhorn/backupstore.CreateDeltaBlockBackup()
      /go/src/github.com/longhorn/backupstore/deltablock.go:221 +0x1452
  github.com/longhorn/backupstore/test.(*TestSuite).createAndWaitForBackup()
      /go/src/github.com/longhorn/backupstore/test/backup_test.go:234 +0x50
  github.com/longhorn/backupstore/test.(*TestSuite).TestBackupBasic()
      /go/src/github.com/longhorn/backupstore/test/backup_test.go:321 +0xb32
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:321 +0xd3
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4

Goroutine 32 (running) created at:
  github.com/longhorn/backupstore.(*FileLock).Refresh()
      /go/src/github.com/longhorn/backupstore/lock.go:146 +0x114
  github.com/longhorn/backupstore.(*FileLock).Lock()
      /go/src/github.com/longhorn/backupstore/lock.go:116 +0x62e
  github.com/longhorn/backupstore.CreateDeltaBlockBackup()
      /go/src/github.com/longhorn/backupstore/deltablock.go:124 +0x290
  github.com/longhorn/backupstore/test.(*TestSuite).createAndWaitForBackup()
      /go/src/github.com/longhorn/backupstore/test/backup_test.go:234 +0x50
  github.com/longhorn/backupstore/test.(*TestSuite).TestBackupBasic()
      /go/src/github.com/longhorn/backupstore/test/backup_test.go:321 +0xb32
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:321 +0xd3
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4
==================
==================
WARNING: DATA RACE
Write at 0x00c0002fc5a0 by goroutine 48:
  github.com/longhorn/backupstore.(*FileLock).Unlock()
      /go/src/github.com/longhorn/backupstore/lock.go:126 +0x102
  github.com/longhorn/backupstore.RestoreDeltaBlockBackup.func2()
      /go/src/github.com/longhorn/backupstore/deltablock.go:488 +0x4be

Previous read at 0x00c0002fc5a0 by goroutine 47:
  github.com/longhorn/backupstore.(*FileLock).Refresh.func1()
      /go/src/github.com/longhorn/backupstore/lock.go:149 +0xb5

Goroutine 48 (running) created at:
  github.com/longhorn/backupstore.RestoreDeltaBlockBackup()
      /go/src/github.com/longhorn/backupstore/deltablock.go:465 +0x1085
  github.com/longhorn/backupstore/test.(*TestSuite).TestBackupBasic()
      /go/src/github.com/longhorn/backupstore/test/backup_test.go:332 +0xcef
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:321 +0xd3
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4

Goroutine 47 (running) created at:
  github.com/longhorn/backupstore.(*FileLock).Refresh()
      /go/src/github.com/longhorn/backupstore/lock.go:146 +0x114
  github.com/longhorn/backupstore.(*FileLock).Lock()
      /go/src/github.com/longhorn/backupstore/lock.go:116 +0x62e
  github.com/longhorn/backupstore.RestoreDeltaBlockBackup()
      /go/src/github.com/longhorn/backupstore/deltablock.go:406 +0x3e3
  github.com/longhorn/backupstore/test.(*TestSuite).TestBackupBasic()
      /go/src/github.com/longhorn/backupstore/test/backup_test.go:332 +0xcef
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:321 +0xd3
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4
==================
time="2020-07-10T18:42:10Z" level=warning msg="File /tmp/backupstore-test374216996/restore-0 for the restore exists, will remove and re-create it"
time="2020-07-10T18:42:23Z" level=warning msg="File /tmp/backupstore-test374216996/restore-1 for the restore exists, will remove and re-create it"
==================
WARNING: DATA RACE
Write at 0x00c00013e610 by goroutine 52:
  github.com/longhorn/backupstore.(*FileLock).Unlock()
      /go/src/github.com/longhorn/backupstore/lock.go:126 +0x102
  github.com/longhorn/backupstore.RestoreDeltaBlockBackupIncrementally.func2()
      /go/src/github.com/longhorn/backupstore/deltablock.go:625 +0x1a9

Previous read at 0x00c00013e610 by goroutine 49:
  github.com/longhorn/backupstore.(*FileLock).Refresh.func1()
      /go/src/github.com/longhorn/backupstore/lock.go:149 +0xb5

Goroutine 52 (running) created at:
  github.com/longhorn/backupstore.RestoreDeltaBlockBackupIncrementally()
      /go/src/github.com/longhorn/backupstore/deltablock.go:608 +0xf71
  github.com/longhorn/backupstore/test.(*TestSuite).TestBackupRestoreExtra()
      /go/src/github.com/longhorn/backupstore/test/backup_test.go:560 +0x31bf
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:321 +0xd3
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4

Goroutine 49 (running) created at:
  github.com/longhorn/backupstore.(*FileLock).Refresh()
      /go/src/github.com/longhorn/backupstore/lock.go:146 +0x114
  github.com/longhorn/backupstore.(*FileLock).Lock()
      /go/src/github.com/longhorn/backupstore/lock.go:116 +0x62e
  github.com/longhorn/backupstore.RestoreDeltaBlockBackupIncrementally()
      /go/src/github.com/longhorn/backupstore/deltablock.go:540 +0x2fc
  github.com/longhorn/backupstore/test.(*TestSuite).TestBackupRestoreExtra()
      /go/src/github.com/longhorn/backupstore/test/backup_test.go:560 +0x31bf
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:321 +0xd3
  gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
  gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
      /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4
```

File: test/backup_test.go [hunk 1]
```diff
@@ -479,6 +510,18 @@ func (s *TestSuite) TestBackupRestoreExtra(c *C) {
 	err = snap4.Close()
 	c.Assert(err, IsNil)
 
+	//delta4 file between snap4 and snap3
+	delta4, err := os.Create("delta4")
+	c.Assert(err, IsNil)
+	_, err = delta4.WriteAt(dataEmpty, 1*blockSize)
+	c.Assert(err, IsNil)
+	_, err = delta4.WriteAt(data, 4*blockSize)
```
Comment (joshimoo) [inline test/backup_test.go:518]: Do we need to include, the dataModified writes from snap4 into delta4?
```
	_, err = snap4.WriteAt(dataModified, 2*blockSize) ?
```

File: test/backup_test.go [hunk 2]
```diff
@@ -519,14 +562,15 @@ func (s *TestSuite) TestBackupRestoreExtra(c *C) {
...
 
-			err = os.Rename(restore, restoreIncre)
-			c.Assert(err, IsNil)
 		} else {
 			c.Assert(err, IsNil)
 
 			s.waitForRestoreCompletion(c, &volume)
-			err = exec.Command("diff", restoreIncre, volume.Snapshots[i].Name).Run()
+			deltaName := "delta" + strconv.Itoa(i)
+			logrus.Warningf("res %v", restoreIncre)
```
Comment (joshimoo) [inline test/backup_test.go:570]: left over?

Comment (boknowswiki) [reply test/backup_test.go:570]: Sorry, good catch.

**Comment by boknowswiki** [comment]
Put issue which changed the backup restore incremental behavior [here #1188](https://github.com/longhorn/longhorn/issues/1188)

**Comment by boknowswiki** [comment]
> There are seems race condition in the `FileLock`. @joshimoo
> 
> ```
> WARNING: DATA RACE
> Write at 0x00c0000c23e0 by goroutine 34:
>   github.com/longhorn/backupstore.(*FileLock).Unlock()
>       /go/src/github.com/longhorn/backupstore/lock.go:126 +0x102
>   github.com/longhorn/backupstore.CreateDeltaBlockBackup.func1()
>       /go/src/github.com/longhorn/backupstore/deltablock.go:230 +0x2cd
> 
> Previous read at 0x00c0000c23e0 by goroutine 32:
>   github.com/longhorn/backupstore.(*FileLock).Refresh.func1()
>       /go/src/github.com/longhorn/backupstore/lock.go:149 +0xb5
> 
> Goroutine 34 (running) created at:
>   github.com/longhorn/backupstore.CreateDeltaBlockBackup()
>       /go/src/github.com/longhorn/backupstore/deltablock.go:221 +0x1452
>   github.com/longhorn/backupstore/test.(*TestSuite).createAndWaitForBackup()
>       /go/src/github.com/longhorn/backupstore/test/backup_test.go:234 +0x50
>   github.com/longhorn/backupstore/test.(*TestSuite).TestBackupBasic()
>       /go/src/github.com/longhorn/backupstore/test/backup_test.go:321 +0xb32
>   runtime.call32()
>       /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
>   reflect.Value.Call()
>       /usr/local/go/src/reflect/value.go:321 +0xd3
>   gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
>   gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4
> 
> Goroutine 32 (running) created at:
>   github.com/longhorn/backupstore.(*FileLock).Refresh()
>       /go/src/github.com/longhorn/backupstore/lock.go:146 +0x114
>   github.com/longhorn/backupstore.(*FileLock).Lock()
>       /go/src/github.com/longhorn/backupstore/lock.go:116 +0x62e
>   github.com/longhorn/backupstore.CreateDeltaBlockBackup()
>       /go/src/github.com/longhorn/backupstore/deltablock.go:124 +0x290
>   github.com/longhorn/backupstore/test.(*TestSuite).createAndWaitForBackup()
>       /go/src/github.com/longhorn/backupstore/test/backup_test.go:234 +0x50
>   github.com/longhorn/backupstore/test.(*TestSuite).TestBackupBasic()
>       /go/src/github.com/longhorn/backupstore/test/backup_test.go:321 +0xb32
>   runtime.call32()
>       /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
>   reflect.Value.Call()
>       /usr/local/go/src/reflect/value.go:321 +0xd3
>   gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
>   gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4
> ==================
> ==================
> WARNING: DATA RACE
> Write at 0x00c0002fc5a0 by goroutine 48:
>   github.com/longhorn/backupstore.(*FileLock).Unlock()
>       /go/src/github.com/longhorn/backupstore/lock.go:126 +0x102
>   github.com/longhorn/backupstore.RestoreDeltaBlockBackup.func2()
>       /go/src/github.com/longhorn/backupstore/deltablock.go:488 +0x4be
> 
> Previous read at 0x00c0002fc5a0 by goroutine 47:
>   github.com/longhorn/backupstore.(*FileLock).Refresh.func1()
>       /go/src/github.com/longhorn/backupstore/lock.go:149 +0xb5
> 
> Goroutine 48 (running) created at:
>   github.com/longhorn/backupstore.RestoreDeltaBlockBackup()
>       /go/src/github.com/longhorn/backupstore/deltablock.go:465 +0x1085
>   github.com/longhorn/backupstore/test.(*TestSuite).TestBackupBasic()
>       /go/src/github.com/longhorn/backupstore/test/backup_test.go:332 +0xcef
>   runtime.call32()
>       /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
>   reflect.Value.Call()
>       /usr/local/go/src/reflect/value.go:321 +0xd3
>   gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
>   gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4
> 
> Goroutine 47 (running) created at:
>   github.com/longhorn/backupstore.(*FileLock).Refresh()
>       /go/src/github.com/longhorn/backupstore/lock.go:146 +0x114
>   github.com/longhorn/backupstore.(*FileLock).Lock()
>       /go/src/github.com/longhorn/backupstore/lock.go:116 +0x62e
>   github.com/longhorn/backupstore.RestoreDeltaBlockBackup()
>       /go/src/github.com/longhorn/backupstore/deltablock.go:406 +0x3e3
>   github.com/longhorn/backupstore/test.(*TestSuite).TestBackupBasic()
>       /go/src/github.com/longhorn/backupstore/test/backup_test.go:332 +0xcef
>   runtime.call32()
>       /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
>   reflect.Value.Call()
>       /usr/local/go/src/reflect/value.go:321 +0xd3
>   gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
>   gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4
> ==================
> time="2020-07-10T18:42:10Z" level=warning msg="File /tmp/backupstore-test374216996/restore-0 for the restore exists, will remove and re-create it"
> time="2020-07-10T18:42:23Z" level=warning msg="File /tmp/backupstore-test374216996/restore-1 for the restore exists, will remove and re-create it"
> ==================
> WARNING: DATA RACE
> Write at 0x00c00013e610 by goroutine 52:
>   github.com/longhorn/backupstore.(*FileLock).Unlock()
>       /go/src/github.com/longhorn/backupstore/lock.go:126 +0x102
>   github.com/longhorn/backupstore.RestoreDeltaBlockBackupIncrementally.func2()
>       /go/src/github.com/longhorn/backupstore/deltablock.go:625 +0x1a9
> 
> Previous read at 0x00c00013e610 by goroutine 49:
>   github.com/longhorn/backupstore.(*FileLock).Refresh.func1()
>       /go/src/github.com/longhorn/backupstore/lock.go:149 +0xb5
> 
> Goroutine 52 (running) created at:
>   github.com/longhorn/backupstore.RestoreDeltaBlockBackupIncrementally()
>       /go/src/github.com/longhorn/backupstore/deltablock.go:608 +0xf71
>   github.com/longhorn/backupstore/test.(*TestSuite).TestBackupRestoreExtra()
>       /go/src/github.com/longhorn/backupstore/test/backup_test.go:560 +0x31bf
>   runtime.call32()
>       /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
>   reflect.Value.Call()
>       /usr/local/go/src/reflect/value.go:321 +0xd3
>   gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
>   gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4
> 
> Goroutine 49 (running) created at:
>   github.com/longhorn/backupstore.(*FileLock).Refresh()
>       /go/src/github.com/longhorn/backupstore/lock.go:146 +0x114
>   github.com/longhorn/backupstore.(*FileLock).Lock()
>       /go/src/github.com/longhorn/backupstore/lock.go:116 +0x62e
>   github.com/longhorn/backupstore.RestoreDeltaBlockBackupIncrementally()
>       /go/src/github.com/longhorn/backupstore/deltablock.go:540 +0x2fc
>   github.com/longhorn/backupstore/test.(*TestSuite).TestBackupRestoreExtra()
>       /go/src/github.com/longhorn/backupstore/test/backup_test.go:560 +0x31bf
>   runtime.call32()
>       /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
>   reflect.Value.Call()
>       /usr/local/go/src/reflect/value.go:321 +0xd3
>   gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:772 +0x9aa
>   gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1()
>       /go/src/github.com/longhorn/backupstore/vendor/gopkg.in/check.v1/check.go:666 +0xc4
> ```

@yasker , how to reproduce this issue?

**Pull request merged**